File: src/main/java/sx/event/EventDispatcher.java
package sx.event;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

/**
 * Multicast event dispatcher abstract and factory.
 * <p/>
 * This class is supposed to be used by observable classes, which can in turn expose instance(s) via EventDelegate interface for consumer to add listeners.
 * Created by masc on 11.08.14.
 */
public abstract class EventDispatcher<T extends EventListener> implements EventDelegate<T> {
    private Logger mLog = LoggerFactory.getLogger(this.getClass());

    /**
     * Listener reference with type information
     */
    class ListenerReference extends WeakReference<T> {
        private Class mClass;

        public ListenerReference(T referent, Class<T> c) {
            super(referent);
            mClass = c;
        }

        public Class getListenerClass() {
            return mClass;
        }
    }

    /**
     * Generic runnable interface for event emission
     *
     * @param <T>
     */
    public interface Runnable<T> {
        void run(T listener);
    }

    public EventDispatcher() {
    }

    /**
     * Add listener
     * @param listener
     */
    public abstract void add(T listener);

    /**
     * Remove listener
     * @param listener
     */
    public abstract void remove(T listener);

    /**
     * Remove weak references from listeners, used for cleaning out garbage collected/invalid rfs
     * @param listeners
     */
    protected abstract void remove(List<ListenerReference> listeners);

    /**
     * Get copy of listener references
     */
    protected abstract List<ListenerReference> getListeners();

    /**
     * Emit event to listeners
     * @param r Runnable supposed to call listener interface method
     */
    public void emit(Runnable<T> r) {
        List<ListenerReference> invalidRefs = null;

        List<ListenerReference> listenerRefs = this.getListeners();
        for (ListenerReference listener : listenerRefs) {
            T instance = listener.get();
            if (instance != null)
                r.run(instance);
            else {
                if (invalidRefs == null)
                    invalidRefs = new ArrayList<ListenerReference>();
                invalidRefs.add(listener);
            }
        }

        // Remove invalid refs
        if (invalidRefs != null) {
            for (ListenerReference ir : invalidRefs) {
                // Convert interface class instances to strings (for logging)
                ArrayList<String> interfaceStrings = new ArrayList<String>();
                for (Class c : ir.getListenerClass().getInterfaces()) {
                    interfaceStrings.add(c.toString());
                }

                mLog.info(String.format("Removing gc'ed listener %s (%s)",
                        ir.getListenerClass(),
                        String.join(",", interfaceStrings)));
            }
            this.remove(invalidRefs);
        }
    }

    /**
     * Factory method to create a regular (thread-unsafe) event dispatcher
     *
     * @param <V> Type of event to dispatch
     * @return Event dispatcher
     */
    public static <V extends EventListener> EventDispatcher<V> create() {
        return new RegularEventDispatcher<V>();
    }

    /**
     * Factory method to create a thread-safe event dispatcher
     *
     * @param <V> Type of event to dispatch
     * @return Event dispatcher
     */
    public static <V extends EventListener> EventDispatcher<V> createThreadSafe() {
        return new ThreadSafeEventDispatcher<V>();
    }
}

File: src/main/java/sx/event/EventDispatcher.kt
package sx.event

import org.slf4j.LoggerFactory

import java.lang.ref.WeakReference
import java.util.ArrayList

/**
 * Multicast event dispatcher abstract and factory.
 *
 *
 * This class is supposed to be used by observable classes, which can in turn expose instance(s) via EventDelegate interface for consumer to add listeners.
 * Created by masc on 11.08.14.
 */
abstract class EventDispatcher<T : EventListener> : EventDelegate<T> {
    private val log = LoggerFactory.getLogger(this.javaClass)

    /**
     * Listener reference with type information
     */
    inner class ListenerReference(referent: T, c: Class<T>) : WeakReference<T>(referent) {
        val listenerClass: Class<*>

        init {
            listenerClass = c
        }
    }

    /**
     * Add listener
     * @param listener
     */
    abstract override fun add(listener: T)

    /**
     * Remove listener
     * @param listener
     */
    abstract override fun remove(listener: T)

    /**
     * Remove weak references from listeners, used for cleaning out garbage collected/invalid rfs
     * @param listeners
     */
    protected abstract fun remove(listeners: List<ListenerReference>)

    /**
     * Get copy of listener references
     */
    protected abstract val listeners: List<ListenerReference>

    /**
     * Emit event to listeners
     * @param r Runnable supposed to call listener interface method
     */
    fun emit(r: (listener: T) -> Unit) {
        var invalidRefs: MutableList<ListenerReference>? = null

        val listenerRefs = this.listeners
        for (listener in listenerRefs) {
            val instance = listener.get()
            if (instance != null)
                r(instance)
            else {
                if (invalidRefs == null)
                    invalidRefs = ArrayList<ListenerReference>()
                invalidRefs.add(listener)
            }
        }

        // Remove invalid refs
        if (invalidRefs != null) {
            for (ir in invalidRefs) {
                // Convert interface class instances to strings (for logging)
                val interfaceListText = ir.listenerClass.interfaces.map { "${it}" }.joinToString(",")

                log.info("Removing gc'ed listener ${ir.listenerClass} (${interfaceListText})")
            }
            this.remove(invalidRefs)
        }
    }

    companion object {
        /**
         * Factory method to create a regular (thread-unsafe) event dispatcher
         * @param <V> Type of event to dispatch
         * @return Event dispatcher
         */
        @JvmStatic fun <V : EventListener> create(): EventDispatcher<V> {
            return RegularEventDispatcher()
        }

        /**
         * Factory method to create a thread-safe event dispatcher
         * @param <V> Type of event to dispatch
         * @return Event dispatcher
         */
        @JvmStatic fun <V : EventListener> createThreadSafe(): EventDispatcher<V> {
            return ThreadSafeEventDispatcher()
        }
    }
}
